### DISCUSSION TITLE
How to chain multiple writes together?

### DISCUSSION THREAD

**Discussion by dapjeongneo** [wevm/viem#837] [Question] [answered]
Hi team,

Is it possible to chain "approve" and "swap" write transaction together? For example, a user wants to swap Token A for Token B. In the normal workflow, it would require the user to click on "approve Token A" then "swap Token A for Token B." 

Is it possible to do something like 
```
writeApprove().then(() => writeSwap())
```
such that the would tx would only require a button click instead of 2? Thanks!

**Comment by saintmaxi** [comment, marked as answer]
you would probably do something like this (assuming you're using a [contract instance](https://viem.sh/docs/contract/getContract.html)), leveraging [waitForTransactionReceipt](https://viem.sh/docs/actions/public/waitForTransactionReceipt.html) to detect when the first transaction completes and fire off the second one based off its status - can make it cleaner with async/await logic

```js
contract.write.approve([...args]).then((hash) => {
  publicClient.waitForTransactionReceipt({ hash: hash }).then((receipt) => {
    if (receipt.status === "success") {
      contract.write.swap([...args]).then((hash) => {
        publicClient.waitForTransactionReceipt({ hash: hash }).then((receipt) => {
          if (receipt.status === "success") {
            // handle both successful
          }
          else {
            // handle swap failure
          }
        })
      })
    } 
    else {
      // handle approve failure
    }
  })
});
```
